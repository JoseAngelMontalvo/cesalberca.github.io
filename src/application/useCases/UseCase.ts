export type UseCase = 'GetArticle' | 'GetAllArticles' | 'GetTalksGiven'
