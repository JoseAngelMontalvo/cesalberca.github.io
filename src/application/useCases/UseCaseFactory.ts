import { UseCase } from './UseCase'
import { GetAllArticles, GetArticle, GetTalksGiven } from './index'
import { ArticlesFileRepository } from '../../infraestructure/articles/ArticlesFileRepository'
import { FileLoader } from '../../infraestructure/FileLoader'
import { TranslationService } from '../TranslationService'
import { Translator } from '../../infraestructure/language'

export class UseCaseFactory {
  public static get<T extends UseCase>(useCase: T, context?: any) {
    switch (useCase) {
      case UseCase.GET_ALL_ARTICLES:
        return new GetAllArticles(
          new ArticlesFileRepository(
            FileLoader.create(),
            TranslationService.create(Translator.create())
          ),
          context.locale
        )
      case UseCase.GET_TALKS_GIVEN:
        return new GetTalksGiven()
      case UseCase.GET_ARTICLE:
        return new GetArticle(
          new ArticlesFileRepository(
            FileLoader.create(),
            TranslationService.create(Translator.create())
          ),
          context.id,
          context.locale
        )
      default:
        throw new Error(`Use case "${useCase}" not found`)
    }
  }
}
