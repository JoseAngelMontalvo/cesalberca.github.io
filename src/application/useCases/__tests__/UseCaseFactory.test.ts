import { UseCaseFactory } from '../UseCaseFactory'
import { UseCase } from '../UseCase'

describe('UseCaseFactory', () => {
  it('should create an instance of all use cases', () => {
    const getArticle = UseCaseFactory.get(UseCase.GET_ARTICLE, { id: 1, locale: 0 })
    const getAllArticles = UseCaseFactory.get(UseCase.GET_ALL_ARTICLES, { locale: 0 })
    const getTalksGiven = UseCaseFactory.get(UseCase.GET_TALKS_GIVEN, { locale: 0 })

    expect(getArticle).toMatchSnapshot()
    expect(getAllArticles).toMatchSnapshot()
    expect(getTalksGiven).toMatchSnapshot()
  })

  it('should throw an error if an use case is not found', () => {
    const nonExistingUseCase = 'fooBarBaz'

    expect(() => {
      UseCaseFactory.get((nonExistingUseCase as unknown) as UseCase)
    }).toThrowError(new Error('Use case "fooBarBaz" not found'))
  })
})
