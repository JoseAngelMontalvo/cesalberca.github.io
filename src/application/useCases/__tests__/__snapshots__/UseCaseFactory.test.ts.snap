// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UseCaseFactory should create an instance of all use cases 1`] = `
LoggerCommandDecorator {
  "decoratedCommand": GetArticle {
    "articlesRepository": ArticlesFileRepository {
      "fileLoader": FileLoader {},
      "translationService": TranslationService {
        "translator": Translator {
          "translations": Array [
            Array [
              1,
              Array [
                Array [
                  "_theme",
                  "Tema",
                ],
                Array [
                  "_language",
                  "Idioma",
                ],
                Array [
                  "_en",
                  "Inglés",
                ],
                Array [
                  "_es",
                  "Español",
                ],
                Array [
                  "_dark",
                  "Oscuro",
                ],
                Array [
                  "_light",
                  "Claro",
                ],
                Array [
                  "home_title",
                  "Inicio",
                ],
                Array [
                  "home_recentArticles",
                  "Artículos recientes",
                ],
                Array [
                  "article_title",
                  "Artículos",
                ],
                Array [
                  "about_title",
                  "Acerca de mí",
                ],
                Array [
                  "about_description",
                  "I'm a young programmer very interested in good practices, architecture, testing and SOLID principles. I work at [Autentia](https://www.autentia.com/) as a frontend developer (Vue, React, TypeScript, JavaScript, HTML, CSS, etc) although I have worked a bit in Java and Android. I'm continuously on the lookout for things to read, learn and [share](https://www.adictosaltrabajo.com/author/calberca/). I've talked at [JSDay Madrid](https://www.youtube.com/watch?v=aNf1Oos0ZB8&t=1s), Codenares 2018 and 2019, [Ritsi](https://www.youtube.com/watch?v=SuykoyC0T74&t=1s) and several internal talks at my company, like [this one](https://www.youtube.com/watch?v=RFQy-ud8Ec4&t=4477s).",
                ],
                Array [
                  "talks_title",
                  "Charlas",
                ],
                Array [
                  "talks_talkTitle",
                  "Title",
                ],
                Array [
                  "talks_talkAbstract",
                  "Descripción",
                ],
                Array [
                  "talks_talkTopics",
                  "Temas",
                ],
                Array [
                  "talks_talkDifficulty",
                  "Dificultad",
                ],
                Array [
                  "talks_talkLength",
                  "Duración",
                ],
                Array [
                  "talks_talkDifficultyAdvanced",
                  "Avanzado",
                ],
                Array [
                  "talks_talkDifficultyIntermediate",
                  "Intermedio",
                ],
                Array [
                  "talks_talkDifficultyEasy",
                  "Iniciado",
                ],
              ],
            ],
            Array [
              0,
              Array [
                Array [
                  "_theme",
                  "Theme",
                ],
                Array [
                  "_language",
                  "Language",
                ],
                Array [
                  "_en",
                  "English",
                ],
                Array [
                  "_es",
                  "Spanish",
                ],
                Array [
                  "_dark",
                  "Dark",
                ],
                Array [
                  "_light",
                  "Light",
                ],
                Array [
                  "home_title",
                  "Home",
                ],
                Array [
                  "home_recentArticles",
                  "Recent articles",
                ],
                Array [
                  "article_title",
                  "Articles",
                ],
                Array [
                  "about_title",
                  "About me",
                ],
                Array [
                  "about_description",
                  "I'm a young programmer very interested in good practices, architecture, testing and SOLID principles. I work at [Autentia](https://www.autentia.com/) as a frontend developer (Vue, React, TypeScript, JavaScript, HTML, CSS, etc) although I have worked a bit in Java and Android. I'm continuously on the lookout for things to read, learn and [share](https://www.adictosaltrabajo.com/author/calberca/). I've talked at [JSDay Madrid](https://www.youtube.com/watch?v=aNf1Oos0ZB8&t=1s), Codenares 2018 and 2019, [Ritsi](https://www.youtube.com/watch?v=SuykoyC0T74&t=1s) and several internal talks at my company, like [this one](https://www.youtube.com/watch?v=RFQy-ud8Ec4&t=4477s).",
                ],
                Array [
                  "talks_title",
                  "Talks",
                ],
                Array [
                  "talks_talkTitle",
                  "Title",
                ],
                Array [
                  "talks_talkAbstract",
                  "Abstract",
                ],
                Array [
                  "talks_talkTopics",
                  "Topics",
                ],
                Array [
                  "talks_talkDifficulty",
                  "Difficulty",
                ],
                Array [
                  "talks_talkLength",
                  "Length",
                ],
                Array [
                  "talks_talkDifficultyAdvanced",
                  "Advanced",
                ],
                Array [
                  "talks_talkDifficultyIntermediate",
                  "Intermediate",
                ],
                Array [
                  "talks_talkDifficultyEasy",
                  "Beginner",
                ],
              ],
            ],
          ],
        },
      },
    },
    "id": 1,
    "locale": 0,
  },
  "logger": Logger {
    "stdout": Object {
      "error": [Function],
      "info": [Function],
      "warn": [Function],
    },
  },
}
`;

exports[`UseCaseFactory should create an instance of all use cases 2`] = `
LoggerCommandDecorator {
  "decoratedCommand": GetAllArticles {
    "articlesRepository": ArticlesFileRepository {
      "fileLoader": FileLoader {},
      "translationService": TranslationService {
        "translator": Translator {
          "translations": Array [
            Array [
              1,
              Array [
                Array [
                  "_theme",
                  "Tema",
                ],
                Array [
                  "_language",
                  "Idioma",
                ],
                Array [
                  "_en",
                  "Inglés",
                ],
                Array [
                  "_es",
                  "Español",
                ],
                Array [
                  "_dark",
                  "Oscuro",
                ],
                Array [
                  "_light",
                  "Claro",
                ],
                Array [
                  "home_title",
                  "Inicio",
                ],
                Array [
                  "home_recentArticles",
                  "Artículos recientes",
                ],
                Array [
                  "article_title",
                  "Artículos",
                ],
                Array [
                  "about_title",
                  "Acerca de mí",
                ],
                Array [
                  "about_description",
                  "I'm a young programmer very interested in good practices, architecture, testing and SOLID principles. I work at [Autentia](https://www.autentia.com/) as a frontend developer (Vue, React, TypeScript, JavaScript, HTML, CSS, etc) although I have worked a bit in Java and Android. I'm continuously on the lookout for things to read, learn and [share](https://www.adictosaltrabajo.com/author/calberca/). I've talked at [JSDay Madrid](https://www.youtube.com/watch?v=aNf1Oos0ZB8&t=1s), Codenares 2018 and 2019, [Ritsi](https://www.youtube.com/watch?v=SuykoyC0T74&t=1s) and several internal talks at my company, like [this one](https://www.youtube.com/watch?v=RFQy-ud8Ec4&t=4477s).",
                ],
                Array [
                  "talks_title",
                  "Charlas",
                ],
                Array [
                  "talks_talkTitle",
                  "Title",
                ],
                Array [
                  "talks_talkAbstract",
                  "Descripción",
                ],
                Array [
                  "talks_talkTopics",
                  "Temas",
                ],
                Array [
                  "talks_talkDifficulty",
                  "Dificultad",
                ],
                Array [
                  "talks_talkLength",
                  "Duración",
                ],
                Array [
                  "talks_talkDifficultyAdvanced",
                  "Avanzado",
                ],
                Array [
                  "talks_talkDifficultyIntermediate",
                  "Intermedio",
                ],
                Array [
                  "talks_talkDifficultyEasy",
                  "Iniciado",
                ],
              ],
            ],
            Array [
              0,
              Array [
                Array [
                  "_theme",
                  "Theme",
                ],
                Array [
                  "_language",
                  "Language",
                ],
                Array [
                  "_en",
                  "English",
                ],
                Array [
                  "_es",
                  "Spanish",
                ],
                Array [
                  "_dark",
                  "Dark",
                ],
                Array [
                  "_light",
                  "Light",
                ],
                Array [
                  "home_title",
                  "Home",
                ],
                Array [
                  "home_recentArticles",
                  "Recent articles",
                ],
                Array [
                  "article_title",
                  "Articles",
                ],
                Array [
                  "about_title",
                  "About me",
                ],
                Array [
                  "about_description",
                  "I'm a young programmer very interested in good practices, architecture, testing and SOLID principles. I work at [Autentia](https://www.autentia.com/) as a frontend developer (Vue, React, TypeScript, JavaScript, HTML, CSS, etc) although I have worked a bit in Java and Android. I'm continuously on the lookout for things to read, learn and [share](https://www.adictosaltrabajo.com/author/calberca/). I've talked at [JSDay Madrid](https://www.youtube.com/watch?v=aNf1Oos0ZB8&t=1s), Codenares 2018 and 2019, [Ritsi](https://www.youtube.com/watch?v=SuykoyC0T74&t=1s) and several internal talks at my company, like [this one](https://www.youtube.com/watch?v=RFQy-ud8Ec4&t=4477s).",
                ],
                Array [
                  "talks_title",
                  "Talks",
                ],
                Array [
                  "talks_talkTitle",
                  "Title",
                ],
                Array [
                  "talks_talkAbstract",
                  "Abstract",
                ],
                Array [
                  "talks_talkTopics",
                  "Topics",
                ],
                Array [
                  "talks_talkDifficulty",
                  "Difficulty",
                ],
                Array [
                  "talks_talkLength",
                  "Length",
                ],
                Array [
                  "talks_talkDifficultyAdvanced",
                  "Advanced",
                ],
                Array [
                  "talks_talkDifficultyIntermediate",
                  "Intermediate",
                ],
                Array [
                  "talks_talkDifficultyEasy",
                  "Beginner",
                ],
              ],
            ],
          ],
        },
      },
    },
    "locale": 0,
  },
  "logger": Logger {
    "stdout": Object {
      "error": [Function],
      "info": [Function],
      "warn": [Function],
    },
  },
}
`;

exports[`UseCaseFactory should create an instance of all use cases 3`] = `
LoggerCommandDecorator {
  "decoratedCommand": GetTalksGiven {
    "locale": 0,
    "talksRepository": TalksFileRepository {
      "fileLoader": FileLoader {},
      "translationService": TranslationService {
        "translator": Translator {
          "translations": Array [
            Array [
              1,
              Array [
                Array [
                  "_theme",
                  "Tema",
                ],
                Array [
                  "_language",
                  "Idioma",
                ],
                Array [
                  "_en",
                  "Inglés",
                ],
                Array [
                  "_es",
                  "Español",
                ],
                Array [
                  "_dark",
                  "Oscuro",
                ],
                Array [
                  "_light",
                  "Claro",
                ],
                Array [
                  "home_title",
                  "Inicio",
                ],
                Array [
                  "home_recentArticles",
                  "Artículos recientes",
                ],
                Array [
                  "article_title",
                  "Artículos",
                ],
                Array [
                  "about_title",
                  "Acerca de mí",
                ],
                Array [
                  "about_description",
                  "I'm a young programmer very interested in good practices, architecture, testing and SOLID principles. I work at [Autentia](https://www.autentia.com/) as a frontend developer (Vue, React, TypeScript, JavaScript, HTML, CSS, etc) although I have worked a bit in Java and Android. I'm continuously on the lookout for things to read, learn and [share](https://www.adictosaltrabajo.com/author/calberca/). I've talked at [JSDay Madrid](https://www.youtube.com/watch?v=aNf1Oos0ZB8&t=1s), Codenares 2018 and 2019, [Ritsi](https://www.youtube.com/watch?v=SuykoyC0T74&t=1s) and several internal talks at my company, like [this one](https://www.youtube.com/watch?v=RFQy-ud8Ec4&t=4477s).",
                ],
                Array [
                  "talks_title",
                  "Charlas",
                ],
                Array [
                  "talks_talkTitle",
                  "Title",
                ],
                Array [
                  "talks_talkAbstract",
                  "Descripción",
                ],
                Array [
                  "talks_talkTopics",
                  "Temas",
                ],
                Array [
                  "talks_talkDifficulty",
                  "Dificultad",
                ],
                Array [
                  "talks_talkLength",
                  "Duración",
                ],
                Array [
                  "talks_talkDifficultyAdvanced",
                  "Avanzado",
                ],
                Array [
                  "talks_talkDifficultyIntermediate",
                  "Intermedio",
                ],
                Array [
                  "talks_talkDifficultyEasy",
                  "Iniciado",
                ],
              ],
            ],
            Array [
              0,
              Array [
                Array [
                  "_theme",
                  "Theme",
                ],
                Array [
                  "_language",
                  "Language",
                ],
                Array [
                  "_en",
                  "English",
                ],
                Array [
                  "_es",
                  "Spanish",
                ],
                Array [
                  "_dark",
                  "Dark",
                ],
                Array [
                  "_light",
                  "Light",
                ],
                Array [
                  "home_title",
                  "Home",
                ],
                Array [
                  "home_recentArticles",
                  "Recent articles",
                ],
                Array [
                  "article_title",
                  "Articles",
                ],
                Array [
                  "about_title",
                  "About me",
                ],
                Array [
                  "about_description",
                  "I'm a young programmer very interested in good practices, architecture, testing and SOLID principles. I work at [Autentia](https://www.autentia.com/) as a frontend developer (Vue, React, TypeScript, JavaScript, HTML, CSS, etc) although I have worked a bit in Java and Android. I'm continuously on the lookout for things to read, learn and [share](https://www.adictosaltrabajo.com/author/calberca/). I've talked at [JSDay Madrid](https://www.youtube.com/watch?v=aNf1Oos0ZB8&t=1s), Codenares 2018 and 2019, [Ritsi](https://www.youtube.com/watch?v=SuykoyC0T74&t=1s) and several internal talks at my company, like [this one](https://www.youtube.com/watch?v=RFQy-ud8Ec4&t=4477s).",
                ],
                Array [
                  "talks_title",
                  "Talks",
                ],
                Array [
                  "talks_talkTitle",
                  "Title",
                ],
                Array [
                  "talks_talkAbstract",
                  "Abstract",
                ],
                Array [
                  "talks_talkTopics",
                  "Topics",
                ],
                Array [
                  "talks_talkDifficulty",
                  "Difficulty",
                ],
                Array [
                  "talks_talkLength",
                  "Length",
                ],
                Array [
                  "talks_talkDifficultyAdvanced",
                  "Advanced",
                ],
                Array [
                  "talks_talkDifficultyIntermediate",
                  "Intermediate",
                ],
                Array [
                  "talks_talkDifficultyEasy",
                  "Beginner",
                ],
              ],
            ],
          ],
        },
      },
    },
  },
  "logger": Logger {
    "stdout": Object {
      "error": [Function],
      "info": [Function],
      "warn": [Function],
    },
  },
}
`;
